packages/opentelemetry-instrumentation-grpc/src/grpc-js/client/patches.ts

```diff
@@ -0,0 +1,104 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { GrpcInstrumentation } from '../../instrumentation';
+import type * as grpcJs from '@grpc/grpc-js';
+import type { PackageDefinition } from '@grpc/grpc-js/build/src/make-client';
+import { getMethodsToWrap, getPatchedClientMethods } from './utils';
+
+type MakeClientConstructorFunction = typeof grpcJs.makeGenericClientConstructor;
+
+/**
+ * Entry point for applying client patches to `grpc.makeClientConstructor(...)` equivalents
+ * @param this GrpcJsPlugin
+ */
+export function patchClient(
+  this: GrpcInstrumentation,
+  grpcClient: typeof grpcJs
+): (original: MakeClientConstructorFunction) => MakeClientConstructorFunction {
+  const insttrumentation = this;
+  return (original: MakeClientConstructorFunction) => {
+    insttrumentation.getLogger().debug('patching client');
+    return function makeClientConstructor(
+      this: typeof grpcJs.Client,
+      methods: grpcJs.ServiceDefinition,
+      serviceName: string,
+      options?: object
+    ) {
+      const client = original.call(this, methods, serviceName, options);
+      insttrumentation
+        .getShimmer()
+        .massWrap<typeof client.prototype, string>(
+          client.prototype,
+          getMethodsToWrap.call(insttrumentation, client, methods),
+          getPatchedClientMethods.call(insttrumentation, grpcClient)
+        );
+      return client;
+    };
+  };
+}
+
+/**
+ * Entry point for client patching for grpc.loadPackageDefinition(...)
+ * @param this - GrpcJsPlugin
+ */
+export function patchLoadPackageDefinition(
+  this: GrpcInstrumentation,
+  grpcClient: typeof grpcJs
+) {
+  return (original: typeof grpcJs.loadPackageDefinition) => {
+    const plugin = this;
+
+    plugin._logger.debug('patching loadPackageDefinition');
+
+    return function patchedLoadPackageDefinition(
+      this: null,
+      packageDef: PackageDefinition
+    ) {
+      const result: grpcJs.GrpcObject = original.call(
+        this,
+        packageDef
+      ) as grpcJs.GrpcObject;
+      _patchLoadedPackage.call(plugin, grpcClient, result);
+      return result;
+    } as typeof grpcJs.loadPackageDefinition;
+  };
+}
+
+/**
+ * Utility function to patch *all* functions loaded through a proto file.
+ * Recursively searches for Client classes and patches all methods, reversing the
+ * parsing done by grpc.loadPackageDefinition
+ * https://github.com/grpc/grpc-node/blob/1d14203c382509c3f36132bd0244c99792cb6601/packages/grpc-js/src/make-client.ts#L200-L217
+ */
+function _patchLoadedPackage(
+  this: GrpcInstrumentation,
+  grpcClient: typeof grpcJs,
+  result: grpcJs.GrpcObject
+): void {
+  Object.values(result).forEach(service => {
+    if (typeof service === 'function') {
+      this.getShimmer().massWrap<typeof service.prototype, string>(
```

Comment:
Maybe we should remove the leading underscores from these methods since they are a part of the public interface? Public in this case not meaning access control public, but meaning we advertise the use of these functions as OK to use by third parties extending the class.

Reply:
We can't access protected properties from there (at least from a typescript perspective) so we are forced to re-expose them as public properties.

Reply:
Protected should be accessible from subclasses. That's the whole point.

Reply:
you just have to pass the method as param to the function
```
function _patchLoadedPackage(
  massWrap,
  grpcClient: typeof grpcJs,
  result: grpcJs.GrpcObject
): void {
  Object.values(result).forEach(service => {
    if (typeof service === 'function') {
      massWrap<typeof service.prototype, string>(
....
```

Reply:
There are inside the `GrpcInstrumentation` class but for some reasons i couldn't access them even when propagating the current `this`

Reply:
>you just have to pass the method as param to the function

That definitely seems like an unideal experience. I think making them public might actually be the best thing to do. As long as they are documented for use by the instrumentation, I think it's ok. They are already a part of the API of the instrumentation so changing them is a breaking change already.

Reply:
you can access the protected only within the class, otherwise you can pass them as I showed you. There are reasons behind that. If you "util" class depends heavily on the functionality from the class you should not create an util class but you should put this method on the class itself. Otherwise you should refactor util classes not to depend on main class - if they are really the util class. Then trying to add unit test will quickly shows that you might have too many dependencies which should trigger an alert, `something is designed wrongly`. But in this case I would simply pass the needed methods as in my example. The refactoring itself can be done later.

Reply:
> If you "util" class depends heavily on the functionality from the class you should not create an util class but you should put this method on the class itself

Well thats work if we don't mind having >1.5k lines of code in a single class, which i think we all agree that we don't want.

> Protected should be accessible from subclasses

However i agree that since i propagate the correct `this` i should be able to access them, i will try to see why its not the case.

Reply:
> > you just have to pass the method as param to the function
> 
> That definitely seems like an unideal experience. I think making them public might actually be the best thing to do. As long as they are documented for use by the instrumentation, I think it's ok. They are already a part of the API of the instrumentation so changing them is a breaking change already.

wrap, unwrap, etc. should not be a public api. The instrumentation methods we have are basically enable / disable. You should not have access for functions that are used internally to patch things.